extensions/SearchAPI/lib/Util.pm

```diff
@@ -0,0 +1,320 @@
+        |
+        \"\"(?!\"")             # empty double quoted string
+        |
+        "(?>(?:(?>[^"\\]+)|""|\\.)*)+"
+                                # anything inside double quotes, ungreedy
+        |
+        `(?>(?:(?>[^`\\]+)|``|\\.)*)+`
+                                # anything inside backticks quotes, ungreedy
+        |
+        '(?>(?:(?>[^'\\]+)|''|\\.)*)+'
+                                # anything inside single quotes, ungreedy.
+        |
+        /\*[\ \t\r\n\S]*?\*/      # C style comments
+        |
+        (?:[\w:@]+(?:\.(?:\w+|\*)?)*)
+                                # words, standard named placeholders, db.table.*, db.*
+        |
+        (?: \$_\$ | \$\d+ | \${1,2} )
+                                # dollar expressions - eg $_$ $3 $$
+        |
+        \n                      # newline
+        |
+        [\t\ ]+                 # any kind of white spaces
+    )
+}smx;
+
+# Some workflow borrowed from
+# https://metacpan.org/pod/DBIx::Placeholder::Named
+sub named_params {
+  my ($query, $params) = @_;
+
+  # Convert SQL statement in series of tokens
+  my @query_tokens = $query =~ m{$tokenize_re}smxg;
+
+  my $prefix_length = length NAMED_PREFIX;
+  my $suffix_length = length NAMED_SUFFIX;
+
+  my @values;
+  foreach my $token (@query_tokens) {
+    if (substr($token, 0, $prefix_length) eq NAMED_PREFIX) {
+      my $multiple_values = 0;
+
+      # Remove NAMED_PREFIX from beginning
+      my $token_stripped = substr $token, $prefix_length;
+
+      # Check to see if placeholder specifies list of values (ends with NAMED_SUFFIX)
+      my $token_length = length $token_stripped;
+      if (
+        substr($token_stripped, $token_length - $suffix_length, $suffix_length) eq
+        NAMED_SUFFIX)
+      {
+
+        # Remove the NAMED_SUFFIX from the end of the stripped token
+        $token_stripped  = substr $token_stripped, 0, $token_length - $suffix_length;
+        $multiple_values = 1;
+      }
+
+      # Look for passed parameter and value(s) for the placeholder detected
+      if (defined $params->{$token_stripped}) {
+        if ($multiple_values && ref $params->{$token_stripped} ne 'ARRAY') {
+          return (undef,
+            "Parameter $token_stripped not found for $token or not an array");
+        }
+        push @values,
+          (
+          $multiple_values ? @{$params->{$token_stripped}} : $params->{$token_stripped});
+      }
+      else {
+        # No need to proceed if parameter value is missing
+        return (undef, "Parameter $token_stripped not found for $token");
+      }
+
+      # Replace current token position with correct number of placeholders
+      if ($multiple_values) {
+        $token = join ',', map {'?'} @{$params->{$token_stripped}};
+      }
+      else {
+        $token = '?';
+      }
+    }
+  }
+
+  # Create new SQL stratement from tokens
+  my $new_query = join '', @query_tokens;
+
+  return ($new_query, \@values);
+}
+
+sub bug_to_hash {
```

Comment:
It is a good point. The OOP setup used by Bugzilla is more or less home grown and does not have some of the cool built in functionality of Python for displaying of an object in different formats. I could had new methods to Bugzilla::Bug, Bugzilla::Flag, and Bugzilla::User such as $bug->to_hash or similar that would give something like this. We also have functions in Bugzilla::WebService::Bug that do the same thing but the code there is heavily integrated with that old style of API (XMLRPC, JSONRPC) and it would take some kludgery to make it work in the newer API framework. Let me try some things.